lib/bitbucket_rest_api/request/oauth2.rb

```diff
@@ -0,0 +1,54 @@
+require 'base64'
+require 'faraday'
+require 'forwardable'
+
+module BitBucket
+  module FaradayMiddleware
+    class OAuth2 < Faraday::Middleware
+      AUTH_HEADER = 'Authorization'.freeze
+
+      attr_reader :options
+
+      extend Forwardable
+
+      def call(env)
+        reset_oauth2_token! if oauth2_token.expired?
+
+        unless oauth2_token.token.to_s.empty?
+          env[:request_headers][AUTH_HEADER] = %(Bearer #{oauth2_token.token})
+        end
+
+        @app.call env
+      end
+
+      def initialize(app, options={})
+        super app
+        @options = options
+      end
+
+      private
+
+      def reset_oauth2_token!
+        @token = nil
+      end
+
+      def oauth2_token
+        @token ||= oauth2_client.client_credentials.get_token
+      end
+
+      def oauth2_client
+        oauth2_client ||= ::OAuth2::Client.new(
+          options[:client_id],
+          options[:client_secret],
+          site: 'https://bitbucket.org/site',
+          authorize_url: '/site/oauth/authorize',
+          token_url: '/site/oauth2/access_token',
+          connection_opts: options[:connection_options]
```

Comment:
This middleware is only used for bitbucket.org so I have hardcoded these variables.

Reply:
will this make `site` appear twice in the url? not sure how the oauth2 client formats the urls, but if it uses the `site:` option as a base_url it looks like it may work out to be `https://bitbucket.org/site/site/oauth/authorize` for example (site + authorize_url). If im wrong then please ignore ;)

Reply:
It looks like the client must have been removing the path part of the url for the site behind the scenes. I went ahead and removed it from my code.

Good 👁 !